File: generate_offline_manifest.js
#!/usr/bin/env node
/**
 * Generate offline-manifest.json for service worker caching.
 * Includes all assets needed for offline functionality.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Extensions to include in the offline manifest
const ASSET_EXTENSIONS = new Set([
    '.webm', '.webp', '.ogg', '.mp3', '.png', '.jpg', '.jpeg',
    '.js', '.css', '.html', '.glb', '.ply', '.json'
]);

// Directories to exclude
const EXCLUDE_DIRS = new Set([
    'node_modules', '.git', '__pycache__', 'dist',
    '.antigravity', 'tools', 'bak', '_backup',
    'old', 'originals', 'hd', '_old_data_files'
]);

// Patterns to exclude from paths
const EXCLUDE_PATTERNS = [
    '_old', '/dist/', '/.git/', '/node_modules/',
    '/bak/', '/_backup/', '/old/', '/originals/',
    '/hd/', '_old_data_files'
];

function shouldExclude(pathStr) {
    return EXCLUDE_PATTERNS.some(pattern => pathStr.includes(pattern));
}

function collectAssets(dir, baseDir = dir, assets = []) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);

        // Skip excluded directories
        if (entry.isDirectory()) {
            if (EXCLUDE_DIRS.has(entry.name)) {
                continue;
            }
            collectAssets(fullPath, baseDir, assets);
        } else {
            // Check if file has a relevant extension
            const ext = path.extname(entry.name).toLowerCase();
            if (ASSET_EXTENSIONS.has(ext)) {
                // Get relative path from base directory
                const relPath = path.relative(baseDir, fullPath);
                let urlPath = '/' + relPath.replace(/\\/g, '/');

                // Skip excluded patterns
                if (shouldExclude(urlPath)) {
                    continue;
                }

                // URL encode special characters (spaces, etc.) but keep slashes
                const parts = urlPath.split('/');
                const encodedParts = parts.map(part => encodeURIComponent(part));
                const encodedUrl = encodedParts.join('/');

                assets.push(encodedUrl);
            }
        }
    }

    return assets;
}

function generateManifest() {
    const rootDir = __dirname;
    const outputFile = path.join(rootDir, 'offline-manifest.json');

    // Core files that must be included
    const coreFiles = [
        '/',
        '/index.html',
        '/game/index.html',
        '/game/',
        '/game/main.js',
        '/game/styles/sub/deck-style.css',
    ];

    console.log('Collecting assets...');
    const assets = collectAssets(rootDir);

    // Combine core files with assets
    let allFiles = [...coreFiles, ...assets];

    // Remove duplicates and sort
    allFiles = [...new Set(allFiles)].sort();

    // Filter out excluded patterns one more time
    allFiles = allFiles.filter(f => !shouldExclude(f));

    // Write to JSON file
    fs.writeFileSync(outputFile, JSON.stringify(allFiles, null, 2), 'utf-8');

    console.log(`✓ Generated ${outputFile}`);
    console.log(`✓ Total files: ${allFiles.length}`);

    // Print summary by extension
    const extensions = {};
    for (const file of allFiles) {
        const ext = path.extname(file).toLowerCase() || 'no-ext';
        extensions[ext] = (extensions[ext] || 0) + 1;
    }

    console.log('\nFiles by type:');
    Object.entries(extensions)
        .sort(([a], [b]) => a.localeCompare(b))
        .forEach(([ext, count]) => {
            console.log(`  ${ext}: ${count}`);
        });
}

generateManifest();

File: generate_offline_manifest.py
#!/usr/bin/env python3
"""
Generate offline-manifest.json for service worker caching.
Includes all assets needed for offline functionality.
"""

import os
import json
from pathlib import Path
from urllib.parse import quote

# Extensions to include in the offline manifest
ASSET_EXTENSIONS = {
    '.webm', '.webp', '.ogg', '.mp3', '.png', '.jpg', '.jpeg',
    '.js', '.css', '.html', '.glb', '.ply', '.json'
}

# Directories to exclude
EXCLUDE_DIRS = {
    'node_modules', '.git', '__pycache__', 'dist', 
    '.antigravity', 'tools', 'bak', '_backup'
}

# Patterns to exclude from paths
EXCLUDE_PATTERNS = ['_old', '/dist/', '/.git/', '/node_modules/']

def should_exclude(path_str):
    """Check if a path should be excluded."""
    for pattern in EXCLUDE_PATTERNS:
        if pattern in path_str:
            return True
    return False

def collect_assets(root_dir):
    """Collect all asset files recursively."""
    assets = []
    root_path = Path(root_dir)
    
    for file_path in root_path.rglob('*'):
        # Skip directories
        if file_path.is_dir():
            continue
        
        # Skip excluded directories
        if any(excluded in file_path.parts for excluded in EXCLUDE_DIRS):
            continue
        
        # Check if file has a relevant extension
        if file_path.suffix.lower() in ASSET_EXTENSIONS:
            # Get relative path from root
            rel_path = file_path.relative_to(root_path)
            url_path = '/' + str(rel_path).replace('\\', '/')
            
            # Skip excluded patterns
            if should_exclude(url_path):
                continue
            
            # URL encode special characters (spaces, etc.)
            parts = url_path.split('/')
            encoded_parts = [quote(part, safe='') for part in parts]
            encoded_url = '/'.join(encoded_parts)
            
            assets.append(encoded_url)
    
    return sorted(set(assets))

def generate_manifest(root_dir, output_file):
    """Generate the offline manifest JSON file."""
    
    # Core files that must be included
    core_files = [
        '/',
        '/index.html',
        '/game/index.html',
        '/game/',
        '/game/main.js',
        '/game/styles/sub/deck-style.css',
    ]
    
    # Collect all assets
    print("Collecting assets...")
    assets = collect_assets(root_dir)
    
    # Combine core files with assets
    all_files = core_files + assets
    
    # Remove duplicates and sort
    all_files = sorted(set(all_files))
    
    # Filter out excluded patterns one more time
    all_files = [f for f in all_files if not should_exclude(f)]
    
    # Write to JSON file
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(all_files, f, indent=2, ensure_ascii=False)
    
    print(f"✓ Generated {output_file}")
    print(f"✓ Total files: {len(all_files)}")
    
    # Print summary by extension
    extensions = {}
    for file in all_files:
        ext = Path(file).suffix.lower() or 'no-ext'
        extensions[ext] = extensions.get(ext, 0) + 1
    
    print("\nFiles by type:")
    for ext, count in sorted(extensions.items()):
        print(f"  {ext}: {count}")

if __name__ == '__main__':
    root_dir = Path(__file__).parent
    output_file = root_dir / 'offline-manifest.json'
    generate_manifest(root_dir, output_file)
